[Chi Xu] iP

tsiyuk opened the pull request:
# Duke
> “Productivity is never an accident. It is always the result of a commitment to excellence, intelligent planning, and focused effort.” 
> – Paul J. Meyer ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ ___FAST___ to use
* user-friendly
* optimized for command line interface 

All you need to do is,

1. download it from [here](https://github.com/tsiyuk/ip/releases).
2. double-click it.
3. add your todos, events and deadlines.
4. let it manage your tasks for you :smiley:

And it is a __FREE__ and __OPEN-SOURCE__ application

Features:

- [x] Managing todos
- [x] Managing events
- [x] Managing deadlines
- [x] Searching
- [ ] GUI (coming soon)
***
If you are new to java, you can use it to practice java and gain some project experience.
Here's the <font color = "#dd0000">`main`</font> method:
```java
public static void main(String[] args) {
     Scanner sc = new Scanner(System.in);
     try {
         TaskList taskList = new TaskList(storage.readFromTaskTxt());
         run(sc, taskList);
     } catch (DukeException e) {
         Ui.showMessage(e.getMessage());
     } finally {
         sc.close();
     }
}
```

Commit:
Add Gradle support

tsiyuk committed:
Add greet, echo and exit

tsiyuk committed:
Add task management for list and add functionality

tsiyuk committed:
Add mark as done

tsiyuk committed:
Add deadline, todo and event

tsiyuk committed:
Add automated text ui testcases

tsiyuk committed:
fix text ui bugs: wrong space

tsiyuk committed:
Add exception handling

tsiyuk committed:
Add delete

tsiyuk committed:
Add enum and refactor the code

tsiyuk committed:
Add document

tsiyuk committed:
Add save functionality

tsiyuk committed:
Add date time for event and deadline

tsiyuk committed:
Add date time for storage

tsiyuk committed:
Refactor the code

tsiyuk committed:
Divide classes into packages

tsiyuk committed:
Add JUnit test for TaskList and Parser

tsiyuk committed:
Create a JAR file

tsiyuk committed:
Add more java doc

tsiyuk committed:
Modify some java doc

tsiyuk committed:
Tweak the code to comply with a coding standard

tsiyuk committed:
Add find action

tsiyuk committed:
Merge conflicts

tsiyuk committed:
Merge conflicts in javadoc

tsiyuk committed:
Merge conflicts

tsiyuk committed:
Refactor Ui

tsiyuk committed:
Add command package and refactor code

tsiyuk committed:
Merge conflicts

tsiyuk committed:
Merge conflicts

tsiyuk committed:
Merge branch 'add-gradle-support'

tsiyuk committed:
Add gradle support and checkstyle

wilburrito commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,170 @@
+     * @param input The given input string.
+     * @return An action based on the given string.
+     */
+    public static Action getAction(String input) {
```
This is really nice. I like it!

wilburrito commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,126 @@
+ */
+public class Ui {
+    /** Text UI: horizontal line. */
+    public static final String HORIZONTAL_LINE = "  -----------------------";
```
I should probably do this for my project as well. Nicely done 💯

wilburrito commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,64 @@
+        this.info = info;
+    }
+
+    @Override
```
perhaps you could write a javadoc for this method?

aishh12 replied:
I agree, public methods should have javadoc comments associated with them as per the coding standard.

wilburrito commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,57 @@
+import duke.Storage;
+import duke.task.TaskList;
+
+public abstract class Command {
```
Nice OOP!

wilburrito commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * This class represents a todo task.
+ */
+public class Todo extends Task {
```
Personally, I treat "To Do" as two words, but this is honestly nitpicky at this point; great work!

wilburrito reviewed (approved):
Great work! LGTM!

aishh12 commented on src/main/java/duke/Action.java:
```diff
@@ -0,0 +1,6 @@
+package duke;
+
+/** The Action enum enumerates all possible actions. */
+public enum Action {
```
I like how you've used enums to restrict the domain of the actions that can be carried out :)

aishh12 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,170 @@
+        }
+        case EVENT: {
+            if (firstWordIndex == -1) {
+                throw new DukeException("The description of an event cannot be empty.");
```
Maybe you want to consider abstracting out the process of throwing the DukeException for the description of a certain task being empty? As seen above, all the cases look for whether firstWordIndex's value is -1, and in which case it'd throw the appropriate exception where the string only differs by the type of task!

aishh12 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,126 @@
+        }
+        System.out.println(Ui.HORIZONTAL_LINE);
+    }
+}
```
I like how you've grouped all the relevant Ui methods appropriately under this class :)

aishh12 reviewed (commented):
Overall, great job done! I enjoyed reading your code and there were only a few suggestions I had to make. All the best :)

tsiyuk committed:
Add naive GUI

tsiyuk committed:
Change duke's avatar

tsiyuk committed:
Command subclasses: add assertions for execute method

Method execute() lacks an assert feature.

This method makes an assumption that every command has
correct corresponding action type.

Ensures that all commands has the correct corresponding
action type.

Let's add assertions to execute() method. Doing so will
help us in the debug session.

tsiyuk committed:
Ui class: remove duplicate code

Ui has a lot of methods to print text Ui, which have many common
behaviors.

The common behaviors among these methods cause code duplication.

Extracting common behaviors like show multilines message and
show single line message.

Let's extract the common behaviors and write new method.

Using more common method can remove duplicate codes.

tsiyuk committed:
Merge pull request #1 from tsiyuk/branch-A-Assertions

Branch-A_assertions

tsiyuk committed:
Merge branch 'master' of https://github.com/tsiyuk/ip

tsiyuk committed:
Merge conflicts from master

tsiyuk committed:
Merge pull request #2 from tsiyuk/branch-A-CodeQuality

Branch-A-CodeQuality

tsiyuk committed:
Merge branch 'master' of https://github.com/tsiyuk/ip

tsiyuk committed:
Add edit functionality

tsiyuk committed:
Add editing date time functionality and clone command

tsiyuk committed:
Merge branch 'branch-C-Update'

tsiyuk committed:
Add CI

tsiyuk committed:
Merge branch 'branch-A-CI'

tsiyuk committed:
Improve GUI

tsiyuk committed:
Merge branch 'branch-A-BetterGUI'

tsiyuk committed:
Enable wrap text

tsiyuk committed:
Change storage path and modify build.gradle

tsiyuk committed:
Merge branch 'fix-storage'

tsiyuk committed:
Add user guide

tsiyuk committed:
Modify javadoc header comment

tsiyuk committed:
Modify user guide

tsiyuk committed:
Modify user guide

tsiyuk committed:
Modify user guide

tsiyuk committed:
Fix test bug

tsiyuk committed:
Merge branch 'master' of https://github.com/tsiyuk/ip
